### PR TITLE
Library concurency refactoring

### PR DESCRIPTION
**Pull request by troyane**
Hi, developers!

I need to get your code review.

Here is my branch _"LibraryConcurencyRefactoring"_ developed as Google Summer of Code 2013 project.

Here you can read about our implemented ideas: http://mixxx.org/wiki/doku.php/lambda_scheme.

You can contact me via GitHub -- https://github.com/troyane, e-mail -- troyan3 at gmail dot com, or at mixxx irc channel -- tro.

P. S. There are some known issues with Rhythmbox playlist. It will be solved soon.

### CODE REVIEW

File: src/dlghidden.h [hunk 1]
```diff
@@ -26,8 +27,10 @@ class DlgHidden : public QWidget, public Ui::DlgHidden, public LibraryView {
     void selectAll();
     void selectionChanged(const QItemSelection&, const QItemSelection&);
 
+  private slots:
+    void slotActivateButtons(bool enable);
```
Comment (MK-42) [inline src/dlghidden.h]: is there a reason to have this (private) slotted? Could be a private member as it was?!

**Comment by MK-42** [comment]
I do not really understand the benefit from using the lambdas for synchronized calls. Is there any or is it just making things consistent? Or to be sure that the code will be executed on main thread?
I've just solved half of the 162 merge-conflicts... need a pause now and I am curious if this will compile afterwards. Many changes last 2 Years... Still unsure if I am experienced enough for this

**Comment by daschuer** [comment]
Hi Markus, thank you for looking at this branch. 

I am sure merging master to this branch at once will not work.

A working merge approach will take some time. A possible solution is to study the git commit trees, and merge e.g. one by one pull request merges from master to this branch.
After each merge, we have to check if there is something to adapt to the new concept and to stay always in a working state.     

Before start, we have to make sure, that we understand the code (I have to re-learn and than I hope I can give you some hints) and we have a common sense that we are still on the right track. 

It also might have issues.

I hope you are not shocked now and happy to jump in here ;-)

Lets start: 

> I do not really understand the benefit from using the lambdas for synchronized calls. Is there any or is it just making things consistent? 

The great benefit of the lambda solution is, that we keep the business logic of the library code nearly untouched. I have to reread the code, but directly from my head, it works like that: 
A original function, with "prepare / query / use" is spitted in place into the three parts by surrounding it with lambdas. The query part is passed by a queue to the library thread and executed there. 
the use part is executed in a slot triggered after the query result is available. 
m_pTrackCollection->callAsync()

In cases we can't use a slot for the result only the query part is passed to the library thread using m_pTrackCollection->callSync()  

Some more infos:
www.mixxx.org/wiki/doku.php/nonblockingdb_status

**Comment by MK-42** [comment]
> A working merge approach will take some time. A possible solution is to study the git commit trees, and merge e.g. one by one pull request merges from master to this branch.
After each merge, we have to check if there is something to adapt to the new concept and to stay always in a working state.

Well.. this is PR 73, we are at 598 now... this will take a long while.
You mean going through the git log from the last commit on master in this branch and merge the next PR into this one, resolve conflicts, change handling, ... and merge till next PR? Would you like to review after each merge before we go on?
(Don't fear, I will not start immediately, will look into the codebase first) 

> The great benefit of the lambda solution is....

I understood so far:
* We want to move all DB-Accesses into a new thread in `TrackCollection`.
* `TrackCollection` has a queue of lambdas which will be executed subsequently
* We use callAsync to queue lambdas to be executed by that thread.
    * We use MainExecuter::callSync() to call a 'callbackSlot' via mainTread if finished
* We use callSync to submit a lambda and wait until that is executed
    * Q: where is the benefit here from direct calling of the lambdas content?
    * Is it, that it is executed in the other thread and not in UI-Thread? But UIThread is also waiting then or not?

95bc90dde324d7930b4df4c26fe068919e120756 is the last PR-merge on master that is already in this branch.

**Comment by daschuer** [comment]
> Well.. this is PR 73, we are at 598 now... this will take a long while.

Oh yes, fortunately most of them are not library related. So hopefully you find reasonable chunks to speed the work up. 

> Would you like to review after each merge before we go on?

No, if you like to do the job. Go with your own pace, I try to review when I find time. 
My main focus is on the 1.12 release though. 

We have also some pending comments of @kain88-de. Maybe you will make you own mind and we find some ideas to simplify the API a bit. 

> ..
> We use MainExecuter::callSync() to call a 'callbackSlot' via mainTread if finished
> ..

All sqlight access has to be done from a single thread. callSync() just blocks the GUI thread until the query result is available, the function containing the query lamda is continued as usual afterwards. 

You are right callSync(), this fails the target to be not blocking, but it helps to keep Mixxx working since one query is executed from an other task all other queries have to be moved to it as well. 
So all that callSync() should be replaced in long term. But it may requires changing the business logic which requires a lot more testing than this will already require. 
So keeping them is help keeping this one in focus to just provide the facilities for later changes.
This PR has already > 2k Lines, which is a pain for reviewers.
  
As far as I remember there is currently no facility which guarantees that all queries are actually moved. See the discussion above.  Maybe you have an idea for it .. a self test somehow.

**Comment by kain88-de** [comment]
Relearning the code those sound nice to me for as start. When we have a clear view of the concept again if would help to decide if we want to merge this branch or start to rewrite it.

@MK-42  Thanks for picking this up. 

But does any of you know what our status is on C++11 compatible compilers? @rryan does the windows build now support C++11?

**Comment by MK-42** [comment]
> As far as I remember there is currently no facility which guarantees that all queries are actually moved. See the discussion above. Maybe you have an idea for it .. a self test somehow.

That is crucial for me - I can't update this one without such a test. I am thinking of some annotation in a comment (like @movedToDBThread or sth. like this) on every call that has been moved to a lambda and creating a tool that will grep through the code and will indicate every function call to a DAO function that needs to be moved and does not have such an annotation. Like that we wouldn't miss any call to such a function. And be forced to update all of them
This is not optimal, as it is only statically analysing the code and itself doesn't prevent any call but indicates it only. Maybe we could integrate this in the UnitTests somehow so that the CI will alert us of such a call.

> @MK-42 Thanks for picking this up.

Well, I am not sure if I can do this. But let's discuss what is possible...

> But does any of you know what our status is on C++11 compatible compilers? @rryan does the windows build now support C++11?

I think @rryan is responsible for the build-servers. If we don't get C++11 in short time, the effort to get this one back up to the master-codebase would be wasted. So is there any roadmap for this?

**Some general thoughts**
* Pro lambda solution used here:
  * no change in buisness-logic needed, only wrap calls by callSync/callAsync
* contra lambda solution:
  * no guarantee that a call to DAOs is done via lambda
    * see my idea of a comment-annotation and some static-analysing above
  * no single entry-point to the library

I think in general this solution is a good one, because it allows to keep the changes little by using callSync and also allows to integrate async-calls.

Maybe it would be easier to create an analysis tool like stated above, test this against the current state of this branch, merge in the general lambda handling to a new branch and update every call that the analysis tool complains about with a CallSync.
At that point we would have not changed the logic in any way, have no unblocking calls but all the calls in the other thread. We could review that (smaller) PR, pull it in and then go on making the calls that are heavy asynchronously. We could speed that second step up by commenting any call that should be made async afterwards in the first runthrough with a todo.

**Comment by daschuer** [comment]
> Test facility ...

> Maybe it would be easier to create an analysis tool like stated above, test this against the current state of this branch, merge in the general lambda handling to a new branch and update every call that the analysis tool complains about with a CallSync.

I do not have a clue about the effeort of a reliable test tool. Maybe it is helpful to you during the development, but I am afraid it will not meet the goal for future changes. 
My hope is that there is a way to use C++ tricks that suits more. 
Like make essential functions private and publish them only to a class accessible by the lambdas. 
I think its hard to be here bulit prove. But I hope it is possible to fail compiling if on tries to access the DB the normal way, or if we just fail to find all instances. If you are a bad guy you will always find a way around such protection. ;-)

Unittest are helpful as well, but these test can  only verify that known code passed are working as desired, and not broken by later commits.

> C++ 

At the time this branch was started there where already c++11 compilers for all targets. Only the Mac environment update was missing. As far as I know today we are on a clang version that supports c++11. Conclusion: there is no reason nnot to use c++11 for Mixxx 1.13.
@rryan: can you verify that? 

>  contra lambda solution:
>>    no single entry-point to the library

The lamdas do not prevent this. I think we can setup a single enty-point based on lambdas that can be used for new code.
Since lamdas are (very) hard to understand it is probably a good idea to provide an API that can be used to access the library without c++11 knowledge.

**Comment by MK-42** [comment]
> Like make essential functions private

I'm not that experienced with lambdas, but I think they can only access methods that the caller can access. You can pass [this] or any other parameter in the lambda and access it inside of it but I doubt there is a possibility to have a function be only callable from a lambda.

We could think of adding a 'Am I executed in the right thread' functionality to all DAO-Functions with database-access which will at least throw a runtime-warning but would mean a check more on every DB-access. But it will compile anyway if someone uses the function without a lamda-move to thread.

**Comment by rryan** [comment]
All build server compilers should support C++11 now. I was going to author a section of the style guide to list which C++11 features we allow though. 1.12 is the last release that will support pre-c++11 and qt4.

I plan to be heavily involved in the library concurrency refactoring. This branch was a good exploration though I don't think we should pre-suppose that the approach it takes is the best one so we should stay open to alternate approaches.

**Comment by MK-42** [comment]
Maybe we should then at first collect which features we want from a new approach and then think of solutions rather than thinking about how to get this here in a mergeble state.

For me it looks like this:
* We want async DB-Calls. Those need to be in another thread.
* Therefore we need to push all DB-Access in this, other, thread.
* We must prevent anyone from calling a DB-Query from another than the DB-Thread
* We want an approach which gives as less changes as possible for now to keep it simple/reviewable?

**Comment by daschuer** [comment]
@MK-42: 
I agree to your list. 
I would like to add: 
* it should be unit testable 

Some times ago I have done some experiments how to hide the normal DB api from outside the Lambdas and the other thread. I have lost the code, but as far as I remember It is possible to pass a pointer to the private class hiding the DB  access from the rest of Mixxx.

**Comment by MK-42** [comment]
> I would like to add:
I've added that to the list. If you have a idea how you've done that, that would be nice. Maybe by some friend-class or sth. like this?

**Comment by MK-42** [comment]
> it should be unit testable

Just thought about it again. What do you want to be testable? If the access to the db is done right or all the individual functions if they do things right?

**Comment by daschuer** [comment]
I original meant that we are able to add uinttests that verifies the new API works as desired.  
Of cause it would also be nice If new API allows to implement unit tests for the individual functions.
But implementing them likely involves a change of the business logic, hat should be avoided in the first step.

**Comment by daschuer** [comment]
I have managed to recall the private DB access idea: 

The lambdas are called here: 
https://github.com/troyane/mixxx/blob/LibraryConcurencyRefactoring/src/library/trackcollection.cpp#L82
with the context, captured in [...] 

Here: https://github.com/troyane/mixxx/blob/LibraryConcurencyRefactoring/src/dlgautodj.cpp#L53
for example we access the database by the captured "this" pointer and the m_pTrackCollection member. 

If we change the lambda prototype to  [...] (TrackCollection* pTrackCollection ) { }
and invoke it by lambda(this) in TrackCollection.
we can remove all direct TrackCollection references from the library access code. 

This dos not entirely hide the TrackCollection functions, but the instance pointer to it.
I think this will fulfill a huge part our requirements. 

In a next step we may split TrackCollection in a way, that the library access functions are divided between GUI thread functions and TrackCollection thread functions. 
Finally a new TrackCollectionDB pointer can be a private Member of a new TrackCollectionGUI class. 
This split is probably the hardest part. 

What do you think?

**Comment by MK-42** [comment]
Thats neat! I would prefer this over the current solution a lot, because the Compiler prevents wrong usage then! So we should go this way i think. That i would not fear that much that we Miss a call somewhere and generate a Party stopper by that.

At the Moment I think it will be easier to Start from current master with this and maybe reuse the trackcollection parts with handling and locking the lambdas. The Main executer would be usable as well maybe.

But we should stay on track with 1.12 release at first. Any point i can help there? Is it only Windows making problems?

**Comment by daschuer** [comment]
We have still a long list of 1.12.0 bugs any help is appreciated.
https://launchpad.net/mixxx/+milestone/1.12.0
However since any software has bugs and Mixxx 1.12 is much better compared to 1.11 IMHO we should push it out the door once we have fixed or moved the critical bugs from the 1.12 milestone.

**Comment by daschuer** [comment]
For this branch: 
It contains the work from a whole summer of code, including many detail fixes.
If we agree, to continue a way like this, It is reasonable to me to go the 800 PR merging pass.
By going this way GIT will axiomatically point you to the regions you have to take hand on. :-) 
You will face conflicts where you have to do work in a new approach as well.  
For my feeling this will be faster than starting again from master. 
And you will hopefully discover bugs, with your pair of unbiased eyes.

**Comment by MK-42** [comment]
I'm looking for a solution to hide the database-access from everywhere except the lambdas.

I like your proposal to make the lambda prototype 
```[...] (TrackCollection* pTrackCollection ) { }```
to prevent anyone calling anything without a lambda.
We would have to make sure that noone uses the TrackCollection pointer without a lambda and everything would be fine.

The TrackCollection instance is already moved to `mixxx.cpp` in this PR. We then need to check if no one passes that instance anywhere and would be sure that no one can access the database without the lambdas.

Are these assumptions correct?

**Comment by troyane** [comment]
Hello, @daschuer @MK-42 @rryan! 
I want to apologize for my absence. But probably all of you heard what is happening now in my country..

I'm happy to hear that this branch is resurrected and someone is still interested in such tricky (but working) approach to this task. It will be pleasant to me if this branch will be merged :) 

I want to help, so I'll try to find several hours each week to help.

**Comment by MK-42** [comment]
Welcome back @troyane !

I noticed that I've missed something in my last comment:
we need the instance of trackCollection to call `call[As|S]ync` on. We can only prevent this by splitting TrackCollection as @daschuer mentioned above.

Would it make sense to do this change right from the start - to prevent wrong usage? This will not make the PR much larger but safer or what do you think?

**Comment by troyane** [comment]
@MK-42 

 - I need to remember all what is going on here. By the way, it can be very useful to you this article: http://mixxx.org/wiki/doku.php/lambda_scheme and this small example of lambdas usage -- https://github.com/troyane/lambdaConcurrent.

 - I do really understand that there will be huge diff because code was written ~ 2 years ago. That's why IMO we need to prepare some adoption. Lets discuss how to do it in easiest way.

 - I've read about splitting of TrackCollection. Unfortunately, I can't understand main reason for this. But, hope we can discuss all.

**Comment by MK-42** [comment]
> I've read about splitting of TrackCollection. Unfortunately, I can't understand main reason for this. But, hope we can discuss all.

The problem is the following:
* We need every call to a dao/Database to be made through a lambda to have it in the right thread
* At the moment we can not guarantee, that we don't miss a point in the code where someone added a `trackCollection->getDao->doSomething` without putting that in a lambda
* We have to manually merge ~400 PRs into this branch to get it up to master again so that will happen a lot.
* Ideally no one can ever call a DB-function from the wrong thread.
The Idea we/@daschuer had was:
* Move everything except `callSync/callAsync` from `TrackCollection` to some other class. Lets call it `TrackCollectionPrivate` for now
* `TrackCollection` holds a private pointer on `TrackCollectionPrivate`
* pass that pointer to each lambda on execution.
That way you could do something like
```
m_pTrackCollection->callSync(
            [&playlistId] (TrackCollectionPrivate pTrackCollectionPrivate) {
        PlaylistDAO& playlistDao = pTrackCollectionPrivate->getPlaylistDAO();
        playlistId = playlistDao.getPlaylistIdFromName(AUTODJ_TABLE);
        if (playlistId < 0) {
            playlistId = playlistDao.createPlaylist(AUTODJ_TABLE,
                                                    PlaylistDAO::PLHT_AUTO_DJ);
        }
    }, __PRETTY_FUNCTION__);
```
what would be called like
```
lambda(m_pTrackCollectionPrivate);
```
but no one except the TrackCollection itself has a reference to `TrackCollectionPrivate` so the methods of it can only be executed out of a lambda in `call[As|S]ync`.

I hope I stated everything right, @daschuer, correct me if I'm wrong.

Maybe it would make sense to then call `TrackCollection`->`Database` and `TrackCollectionPrivate`->`TrackCollection` again.

**Comment by MK-42** [comment]
Any opinions on splitting these two classes now in the current state of the branch and then merging against master? That would make sure (i think) that no code can call the DB the wrong way.
It would make the diff a little bigger though but I think only in places where we have changes anyway.

**Comment by daschuer** [comment]
@troyane: I am happy you are still taking care for your baby. Thank you. 

@MK-42: sorry for the late responds. I think you got the issue correct. 

> Any opinions on splitting these two classes now in the current state of the branch and then merging against master?

This sounds reasonable, if there are not much changes in TrackCollection since splitting out this branch. 
Normally it makes more sense to introduce new changes in top of a merged branch but in this case, we need a proof of concept without wasting time for merge before. 

I am afraid doing the entire split will be quite hard, because of all the criss cross dependencies inside  
TrackCollection. Maybe it is reasonable to start with TrackCollectionPrivate and TrackCollection
which are friend classed from each others. This way you are able to cut one dependency by an other while the code still compiles and works.

**Comment by MK-42** [comment]
>  if there are not much changes in TrackCollection since splitting out this branch

There was some code added to introduce sorting and like in SQLite3 I think. But no major changes except from that one.

I'm working on the Split of these classes right now and making good progress... The Library-Scanner will be tricky though

**Comment by MK-42** [comment]
my work of the last day....

https://github.com/MK-42/mixxx/tree/LibraryConcurencyRefactoring

The LibraryScanner has some direct references to the db and uses its own daos... maybe @troyane or @daschuer can explain how this actually works so that I can bring in the split there too?

nervous to see if this will work :)

**Comment by daschuer** [comment]
If you do a PR to this branch, we can use the nice GitHub features. 
We have to decide what are the best merging steps though.  
@troyane: any preferences?

**Comment by MK-42** [comment]
just opened #629 for your @daschuer

**Comment by rryan** [comment]
Superseded by #629.

**Pull request closed without merging**